spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,319 @@
+External readers can provide any number of resource and/or module schemes and will be configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+[source,pkl]
+----
+/// May be specified as an absolute path to an executable
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+For security reasons, schemes provided by external readers are not allowed by default and clients must explicitly allow the desired scheme(s).
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+If a codebase or user wishes to enforce that only secure LDAP connections are made, the `ldaps:` scheme may be allowed while `ldap:` is not.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+=== Overall Flow
+
+1. User requests an evaluator (via Java API, Message Passing API, or CLI) with an external reader.
+2. Entrypoint (client Java code, `CliEvaluator`, or `Server`) instantiates an `ExternalReader`.
+3. The `ExternalReader` is started, which starts the configured child process and sends a `DiscoverReadersRequest` message.
+4. A `DiscoverReadersResponse` message is received (within some timeout).
+5. The entrypoint requests resource readers (`ExternalResourceReader`) and module key factories (`ExternalModuleKeyFactory`) from the `ExternalReader` and uses them to build an evaluator.
+`ExternalResourceReader` and `ExternalModuleKeyFactory` function similarly to the existing `ClientResourceReader` and `ClientModuleKeyFactory`, implementation may be reusable entirely.
+6. Evaluation begins.
+7. A resource or module read is encountered.
+8. The appropriate `Read*Request` message is sent to the child process, the corresponding `Read*Response` message is awaited.
```

Comment:
In our evaluator API, the order of readers matters; when reading a module, Pkl will try the first reader, then the second, then the third, and so on. In Java, and in jpkl, the last reader is `GenericUrl`, which tries to open all URIs as URLs (for example, it can handle `ftp`).

If an external reader is provided, what is its precedence compared to other readers?

Reply:
That's a great question!

Do you think it would be reasonable to do something like this? From highest to lowest precedence:
1. Any pkl-server `Client` readers - these are currently first and it seems important to preserve this for compatibility
2. Any external readers configured via CLI / Message passing API, in order specified
3. Any external readers configured via PklProject, in order specified
4. All of Pkl's built-in readers in the current order

The current behavior of pkl-server client readers actually surprised me a little.
If there's concern around overriding Pkl's built-in readers, it might make sense to instead do something like this:
1. Any pkl-server `Client` readers
2. All of Pkl's built-in readers in the current order, except `GenericUrl`
3. Any external readers configured via CLI / Message passing API, in order specified
4. Any external readers configured via PklProject, in order specified
5. `GenericUrl`

Reply:
> 1. Any pkl-server `Client` readers - these are currently first and it seems important to preserve this for compatibility
> 2. Any external readers configured via CLI / Message passing API, in order specified
> 3. Any external readers configured via PklProject, in order specified
> 4. All of Pkl's built-in readers in the current order

This makes sense to me! But note that the evaluator doesn't actually about `PklProject` is. `CreateEvaluatorRequest` expects that the client that interacts with to evaluate a PklProject beforehand, and then make its own decisions about what to do with the evaluator settings found there. Our various language bindings have methods that mimick the precedence used by pkl-cli, which is similar to what you've laid out here.

Also: we should consider adding this to `~/.pkl/settings` too.
This enables the use-case of installing a reader globally, which then enables Pkl to use that reader anywhere.

> If there's concern around overriding Pkl's built-in readers, it might make sense to instead do something like this:

We don't recommend doing this, but Pkl intentionally allows this for those that really need to monkey patch Pkl.

Reply:
One advantage of the second approach (combined with the `ResourceReaderFactory` approach from https://github.com/apple/pkl/pull/647) is that no external reader process can be spawned unless a read is attempted for a protocol that Pkl does not implement its own reader native for.

> Also: we should consider adding this to ~/.pkl/settings too.
> This enables the use-case of installing a reader globally, which then enables Pkl to use that reader anywhere.

I had explicitly omitted that from this proposal as I couldn't think of a use case for it. Do you have a use case in mind for this?

Reply:
> Pkl intentionally allows this for those that really need to monkey patch Pkl

This might be an interesting route toward implementing authentication for HTTP(S) resources in lieu of specific handling for this in the language!